### PR TITLE
Adds microbenchmarking and uses standard 1e9 for memory calcs.

### PR DESCRIPTION
**Pull request by patemotter**
Will need to add a smoke test appended to this pr or in the next pr when adding/changing the benchmark.

### CODE REVIEW

File: MaxText/decode.py [hunk 1]
```diff
@@ -23,6 +23,9 @@
 import pyconfig
 import sys
 
+from jax.experimental.compilation_cache import compilation_cache as cc
```
Comment (rwitten) [inline MaxText/decode.py:26]: Can you put this into pyconfig as we discussed?

File: MaxText/microbenchmark.py [hunk 2]
```diff
@@ -0,0 +1,184 @@
+"""Tests for a mock version of the engine API, used in integration tests elsewhere.
```
Comment (rwitten) [inline MaxText/microbenchmark.py:1]: inference_microbenchmark? I think?

Comment (rwitten) [reply MaxText/microbenchmark.py:1]: Also this prefix will cause problems (and is wrong).

File: MaxText/microbenchmark.py [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+
+  ar_results = ar_benchmark(config, engine, params, steps = num_steps, cache_size=cache_size, model_size=model_size)
+
+  for prefill_length in prefill_lengths:
+    tokens, true_length = token_utils.tokenize_and_pad(text, vocab, is_bos=True, prefill_lengths=[prefill_length])
+    prefill_results = prefill_benchmark(config, engine, params, tokens, true_length, num_steps, num_model_params)
+    prefill_results.update(ar_results)
+    results[prefill_length] = prefill_results
+
+  with open("benchmark_results.json", "w") as f:
```
Comment (rwitten) [inline MaxText/microbenchmark.py:170]: can you make sure that sharegpt analysis can consume this? (And you can remove idealization from the sharegpt analysis.)

File: MaxText/microbenchmark.py [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+    json.dump(results, f, indent=2)
+
+
+if __name__ == "__main__":
+  jax.config.update('jax_default_prng_impl', 'unsafe_rbg')
+  os.environ["TF_CPP_MIN_LOG_LEVEL"] = "0"
+  os.environ["M_ICI_FSDP_PARALLELISM"] = "1"
+  os.environ["M_ICI_AUTOREGRESSIVE_PARALLELISM"] = "-1"
```
Comment (rwitten) [inline MaxText/microbenchmark.py:178]: pretty intense. I'd like you to hold off on this. In the future we might want different settings.

File: run_microbenchmark.sh [hunk 5]
```diff
@@ -0,0 +1,21 @@
+#export M_LOAD_PARAMETERS_PATH=gs://maxtext-gamma/7b/2024-02-29-17-56/0/default
+# export TPU_STDERR_LOG_LEVEL=0
+# export TPU_MIN_LOG_LEVEL=0
+# export TF_CPP_MIN_LOG_LEVEL=0
+export M_ICI_FSDP_PARALLELISM=1
+export M_ICI_AUTOREGRESSIVE_PARALLELISM=-1
```
Comment (rwitten) [inline run_microbenchmark.sh:6]: can we put this into the inference scratch world?

**Review by rwitten — CHANGES_REQUESTED** [review]
Some nits!

P.S. - for faster responses in the future, please mark me as having next steps!

**Comment by patemotter** [comment]
1. Improved readability of benchmark output. 
2. Now prints two lines that you can either copy and paste into sharegpt_analyze.py or you can load the json output directly. 
3. Refactored the analyze_sharegpt.py script. 
4. Moved cache compilation method out of individual scripts and into pyconfig.

**Comment by patemotter** [comment]
Removed unnecessary engine inits.

**Pull request closed without merging**